File: mcp/src/lib/units.ts
/** Observations, not a guarantee about future EasyEDA versions. Verify against known geometry. */
export const OBSERVED_PCB_UNITS = Object.freeze({ primitiveMmPerUnit: 0.0254, pouredMmPerUnit: 0.254 });
function finite(value: number, name: string): number {
    if (typeof value !== 'number' || !Number.isFinite(value)) throw new TypeError(`${name} must be a finite number`);
    return value;
}
/** Convert a length only. Never apply blindly to polygon tokens, angles or enum values. */
export function convertLength(value: number, sourceMmPerUnit: number, targetMmPerUnit = 1): number {
    finite(value, 'value'); finite(sourceMmPerUnit, 'sourceMmPerUnit'); finite(targetMmPerUnit, 'targetMmPerUnit');
    if (sourceMmPerUnit <= 0 || targetMmPerUnit <= 0) throw new RangeError('Unit scales must be positive');
    return finite(value * sourceMmPerUnit / targetMmPerUnit, 'converted length');
}
export function milToMm(value: number): number { return convertLength(value, 0.0254); }
export function mmToMil(value: number): number { return convertLength(value, 1, 0.0254); }
/** Read-only numeric sanity check; referenceMm must be independently known and nonzero. */
export function assertUnitScale(nativeDistance: number, referenceMm: number, mmPerUnit: number, toleranceMm: number): void {
    finite(referenceMm, 'referenceMm'); finite(toleranceMm, 'toleranceMm');
    if (nativeDistance <= 0 || referenceMm <= 0 || toleranceMm < 0 || toleranceMm >= referenceMm / 2) throw new RangeError('Use positive distances and a meaningful nonnegative tolerance');
    const actual = convertLength(nativeDistance, mmPerUnit);
    if (Math.abs(actual - referenceMm) > toleranceMm) throw new Error(`Unit scale mismatch: observed ${actual} mm; expected ${referenceMm} ± ${toleranceMm} mm. Recheck the connected EasyEDA API.`);
}

File: mcp/python/easyeda_copilot/units.py
"""Length utilities. Observed API scales require a basic check on each new editor version."""
import math
from types import MappingProxyType

OBSERVED_PCB_UNITS = MappingProxyType({'primitive_mm_per_unit': .0254, 'poured_mm_per_unit': .254})

def _finite(value, name):
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value):
        raise ValueError(f'{name} must be a finite number')
    return value

def convert_length(value, source_mm_per_unit, target_mm_per_unit=1):
    """Lengths only: polygon arrays also contain angles and command parameters."""
    _finite(value, 'value'); _finite(source_mm_per_unit, 'source_mm_per_unit'); _finite(target_mm_per_unit, 'target_mm_per_unit')
    if source_mm_per_unit <= 0 or target_mm_per_unit <= 0:
        raise ValueError('Unit scales must be positive')
    return _finite(value * source_mm_per_unit / target_mm_per_unit, 'converted length')

def mil_to_mm(value): return convert_length(value, .0254)
def mm_to_mil(value): return convert_length(value, 1, .0254)

def assert_unit_scale(native_distance, reference_mm, mm_per_unit, tolerance_mm):
    """Reference length must be independently known; this does not query or modify the editor."""
    _finite(native_distance, 'native_distance'); _finite(reference_mm, 'reference_mm'); _finite(tolerance_mm, 'tolerance_mm')
    if native_distance <= 0 or reference_mm <= 0 or tolerance_mm < 0 or tolerance_mm >= reference_mm / 2:
        raise ValueError('Use positive distances and a meaningful nonnegative tolerance')
    actual = convert_length(native_distance, mm_per_unit)
    if abs(actual-reference_mm) > tolerance_mm:
        raise ValueError(f'Unit scale mismatch: observed {actual} mm; expected {reference_mm} ± {tolerance_mm} mm. Recheck the connected EasyEDA API.')
